src/mapblock_mesh.cpp

```diff
@@ -1083,12 +1104,14 @@ MapBlockMesh::MapBlockMesh(MeshMakeData *data, v3s16 camera_offset):
 			const u16 *indices_p = indices;
 
 			/*
-				Revert triangles for nicer looking gradient if vertices
-				1 and 3 have same color or 0 and 2 have different color.
-				getRed() is the day color.
+				Revert triangles for nicer looking gradient if the
+				brightness of vertices 1 and 3 differ less than
+				the brightness of vertices 0 and 2.
 			*/
-			if(f.vertices[0].Color.getRed() != f.vertices[2].Color.getRed()
-					|| f.vertices[1].Color.getRed() == f.vertices[3].Color.getRed())
+			if (abs(f.vertices[0].Color.getAverage()
+					- f.vertices[2].Color.getAverage())
+					> abs(f.vertices[1].Color.getAverage()
+					- f.vertices[3].Color.getAverage()))
```

Comment:
The amount of sunlight is no longer represented as the red component. The average color is proportional with the brightness of the incoming light.